pkg/property/property.go

```diff
@@ -0,0 +1,322 @@
+	Channels         []Channel
+	GVKs             []GVK
+	GVKsRequired     []GVKRequired
+	Skips            []Skips
+	SkipRanges       []SkipRange
+	BundleObjects    []BundleObject
+
+	Others []Property
+}
+
+const (
+	TypePackage         = "olm.package"
+	TypePackageRequired = "olm.package.required"
+	TypeChannel         = "olm.channel"
+	TypeGVK             = "olm.gvk"
+	TypeGVKRequired     = "olm.gvk.required"
+	TypeSkips           = "olm.skips"
+	TypeSkipRange       = "olm.skipRange"
+	TypeBundleObject    = "olm.bundle.object"
+)
+
+func Parse(in []Property) (*Properties, error) {
+	var out Properties
+	for i, prop := range in {
+		switch prop.Type {
+		case TypePackage:
+			var p Package
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.Packages = append(out.Packages, p)
+		case TypePackageRequired:
+			var p PackageRequired
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.PackagesRequired = append(out.PackagesRequired, p)
+		case TypeChannel:
+			var p Channel
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.Channels = append(out.Channels, p)
+		case TypeGVK:
+			var p GVK
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.GVKs = append(out.GVKs, p)
+		case TypeGVKRequired:
+			var p GVKRequired
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.GVKsRequired = append(out.GVKsRequired, p)
+		case TypeSkips:
+			var p Skips
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.Skips = append(out.Skips, p)
+		case TypeSkipRange:
+			var p SkipRange
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.SkipRanges = append(out.SkipRanges, p)
+		case TypeBundleObject:
+			var p BundleObject
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.BundleObjects = append(out.BundleObjects, p)
+		default:
+			var p json.RawMessage
+			if err := json.Unmarshal(prop.Value, &p); err != nil {
+				return nil, ParseError{Idx: i, Typ: prop.Type, Err: err}
+			}
+			out.Others = append(out.Others, prop)
+		}
+	}
+	return &out, nil
+}
+
+func Deduplicate(in []Property) []Property {
+	type key struct {
+		typ   string
+		value string
+	}
+
+	props := map[key]Property{}
+	var out []Property
+	for _, p := range in {
+		k := key{p.Type, string(p.Value)}
+		if _, ok := props[k]; ok {
+			continue
+		}
+		props[k] = p
+		out = append(out, p)
+	}
+	return out
+}
+
+func ValidateBackCompat(in Properties) error {
```

Comment:
Can we comment on what are the criteria we are using here for backward compatibility?

Reply:
This is a bit of a relic function from 10s of commits ago and as you can see it's much shorter now. I'll revisit this to see if its necessary to keep the last remaining check separated out in this function.